naga/src/back/hlsl/help.rs

```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
                         // End of function body
                         writeln!(self.out, "}}")?;
                     }
+                    crate::MathFunction::Abs
+                        if matches!(
+                            func_ctx.resolve_type(arg, &module.types).scalar(),
+                            Some(crate::Scalar {
+                                kind: crate::ScalarKind::Sint,
+                                width: 4,
+                            })
+                        ) =>
+                    {
+                        let arg_ty = func_ctx.resolve_type(arg, &module.types);
+                        let scalar = arg_ty.scalar().unwrap();
+                        let components = arg_ty.components();
+
+                        let wrapped = WrappedMath {
+                            fun,
+                            scalar,
+                            components,
+                        };
+
+                        if !self.wrapped.math.insert(wrapped) {
+                            continue;
+                        }
+
+                        self.write_value_type(module, arg_ty)?;
+                        write!(self.out, " {ABS_FUNCTION}(")?;
+                        self.write_value_type(module, arg_ty)?;
+                        writeln!(self.out, " val) {{")?;
+
+                        let level = crate::back::Level(1);
+                        writeln!(
+                            self.out,
+                            "{level}return val >= 0 ? val : asint(~asuint(val) + 1u);"
```

Comment:
Doesn't HLSL support unary `-` on `uint`? C++ does. So this could just be 
```suggestion
                            "{level}return val >= 0 ? val : asint(-asuint(val));"
```

Reply:
ah unsigned negation never even occurred to me. Can't see anything in the respective specs, but the shaders appear to pass validation so presumably it's supported.